src/tests/Equinor.ProCoSys.Auth.Tests/Permission/MainApiPermissionServiceTests.cs

```diff
@@ -0,0 +1,211 @@
+        private MainApiPermissionService _dut;
+        private Mock<IMainApiAuthenticator> _mainApiTokenProvider;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _mainApiOptions = new Mock<IOptionsMonitor<MainApiOptions>>();
+            _mainApiOptions
+                .Setup(x => x.CurrentValue)
+                .Returns(new MainApiOptions { ApiVersion = "4.0", BaseAddress = "http://example.com" });
+            _mainApiClient = new Mock<IMainApiClient>();
+
+            _mainApiTokenProvider = new Mock<IMainApiAuthenticator>();
+            _dut = new MainApiPermissionService(_mainApiTokenProvider.Object, _mainApiClient.Object, _mainApiOptions.Object);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldReturnCorrectNumberOfPlants()
+        {
+            // Arange
+            _mainApiClient
+                .Setup(x => x.QueryAndDeserializeAsync<List<AccessablePlant>>(It.IsAny<string>(), null))
+                .Returns(Task.FromResult(new List<AccessablePlant>
+                {
+                    new AccessablePlant { Id = "PCS$ASGARD", Title = "Åsgard" },
+                    new AccessablePlant { Id = "PCS$ASGARD_A", Title = "ÅsgardA" },
+                    new AccessablePlant { Id = "PCS$ASGARD_B", Title = "ÅsgardB" },
+                }));
+
+            // Act
+            var result = await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            Assert.AreEqual(3, result.Count);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldSetsCorrectProperties()
+        {
+            // Arange
+            var proCoSysPlant = new AccessablePlant { Id = "PCS$ASGARD_B", Title = "ÅsgardB" };
+            _mainApiClient
+                .Setup(x => x.QueryAndDeserializeAsync<List<AccessablePlant>>(It.IsAny<string>(), null))
+                .Returns(Task.FromResult(new List<AccessablePlant>
+                {
+                    proCoSysPlant,
+                }));
+            // Act
+            var result = await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            var plant = result.Single();
+            Assert.AreEqual(proCoSysPlant.Id, plant.Id);
+            Assert.AreEqual(proCoSysPlant.Title, plant.Title);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldSetApplicationAuthentication()
+        {
+            // Act
+            await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            _mainApiTokenProvider.VerifySet(a => a.AuthenticationType = AuthenticationType.AsApplication);
+        }
+
+        [TestMethod]
+        public async Task GetAllPlants_ShouldResetToOnBehalfOfAuthentication()
+        {
+            // Act
+            await _dut.GetAllPlantsForUserAsync(_azureOid);
+
+            // Assert
+            _mainApiTokenProvider.VerifySet(a => a.AuthenticationType = AuthenticationType.OnBehalfOf);
+        }
```

Comment:
I think it would be better to use
Assert.AreEqual(AuthenticationType.OnBehalfOf, _mainApiTokenProvider.Object.AuthenticationType);
as VerifySet does not provide any guarantees that it is the last operation performed, and we don't know the state when it exists the GetAllPlantsForUserAsync method.